src/Okta.Idx.Sdk/IdxClient.cs

```diff
@@ -237,6 +240,78 @@ internal async Task<IIdxResponse> IntrospectAsync(IIdxContext idxContext, Cancel
                 request, cancellationToken).ConfigureAwait(false);
         }
 
+        /// <inheritdoc/>
+        public async Task<WidgetSignInResponse> StartWidgetSignInAsync(CancellationToken cancellationToken = default)
+        {
+            var idxContext = await this.InteractAsync();
+            return new WidgetSignInResponse
+            {
+                IdxContext = idxContext,
+                SignInWidgetConfiguration = new SignInWidgetConfiguration(this.Configuration, idxContext),
+            };
+        }
+
+        /// <inheritdoc/>
+        public async Task<TokenResponse> RedeemInteractionCodeAsync(IIdxContext idxContext, string interactionCode, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                Uri issuerUri = new Uri(Configuration.Issuer);
+                Uri tokenUri = new Uri(IdxUrlHelper.GetNormalizedUriString(issuerUri.ToString(), "v1/token"));
+                HttpRequestMessage requestMessage = new HttpRequestMessage(HttpMethod.Post, tokenUri);
+
+                StringBuilder requestContent = new StringBuilder();
+                this.AddParameter(requestContent, "grant_type", "interaction_code", false);
+                this.AddParameter(requestContent, "client_id", Configuration.ClientId, true);
+                if (!string.IsNullOrEmpty(Configuration.ClientSecret))
+                {
+                    this.AddParameter(requestContent, "client_secret", Configuration.ClientSecret, true);
+                }
+
+                this.AddParameter(requestContent, "interaction_code", interactionCode, true);
+                this.AddParameter(requestContent, "code_verifier", idxContext.CodeVerifier, true);
+
+                requestMessage.Content = new StringContent(requestContent.ToString(), Encoding.UTF8, "application/x-www-form-urlencoded");
+                requestMessage.Headers.Add("Accept", "application/json");
+                HttpResponseMessage responseMessage = await this.httpClient.SendAsync(requestMessage);
+                string tokenResponseJson = await responseMessage.Content.ReadAsStringAsync();
+
+                if (!responseMessage.IsSuccessStatusCode)
+                {
+                    throw new RedeemInteractionCodeException(tokenResponseJson);
+                }
+
+                Dictionary<string, object> data = JsonConvert.DeserializeObject<Dictionary<string, object>>(tokenResponseJson);
+                TokenResponse response = new TokenResponse();
+                response.Initialize(this, null, data, _logger);
+                return response;
+            }
+            catch (Exception exception)
+            {
+                LogError(exception);
+                throw new RedeemInteractionCodeException(exception);
+            }
```

Comment:
I'm curious what kind of exceptions could be thrown from the related `try` statement.  Could we have a catch block here for each type of exception?  Thinking from some PHP code that would do this, I would have:
```
try { ... }
catch ($e /InvalidArgumentException) {}
catch ($e {SomeHttpRequestException}) {}
catch ($e {SomeValidationException}) {}
catch ($e /Exception) {//this catches anything not listed above and would be a "terminal" state possibly}
```

Not saying anything is wrong with what is here, just curious if the multiple catch statements could give us and developers any more context into what happens

Reply:
The approach you describe could be potentially helpful; the challenge would be to know what those potential errors are ahead of time to catch specific exception types.  As of now this is a catch-all since it isn't clear what exception(s) may occur, it gets wrapped as RedeemInteractionCodeException so any upstream catchers will at least have the context that it occurred in the RedeemInteractionCode method.